### ISSUE TITLE
[question] What level of testing is "conan test" intended to verify?

### ISSUE THREAD

**Issue by tjwrona** [conan-io/conan#6640]
I am trying to design a library that I wish to create a Conan package from and I'm slightly confused as to what `conan test` is supposed to verify. Is `conan test` supposed to run all of my automated tests for the library (like my "GTest" tests that verify the full library functionality)? Or is it simply supposed to verify that the package can be installed correctly and can be consumed by another project and nothing more?

- [X] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by Minimonium** [comment]
The purpose of `test_package` and by extension of the `conan test` is to verify that the package is packaged correctly. There are could be different ideas on how it _should_ be, but right now it is what it is. 
To run a library's own tests - use the `build` function in the recipe:
https://docs.conan.io/en/latest/reference/conanfile/attributes.html#should-configure-should-build-should-install-should-test

**Comment by memsharded** [comment]
Yes, exactly. The ``test_package`` functionality is intended to test the package, not to test the code or the functionality. So:
- To test that the headers are correctly packaged
- To test that the libraries are correctly packaged
- To test that the runtime or shared libraries are packaged.
- That the binary is compatible, didn't build for the wrong arch or something.
For that, using the package, including and linking the library, and launching a simple app would be enough to catch the majority of problems with the packaging. No testing framework needed.

Besides, even if we tried, it is not possible to run unit tests with this functionality, as it uses the final package. Unit tests many time test internal implementation details of the libraries, that are simply not exposed by the final package.

The correct place to run unit test for the library is the ``build()`` step, which means ``build()`` the package, not only compile the code. If unit-tests fail there, then there is no package at all, then no risk to upload something that is broken. A package that fails in ``test_package`` could still be uploaded if the pipeline doesn't correctly capture the output.

**Comment by Minimonium** [comment]
Hey, @memsharded , sorry to hijack the discussion a bit, but since we are on the topic of `test_package` I want to ask - the `test_package` runs only on recipe creation, usually only for a given configuration. But Conan allows changing the configuration and package layout based on that in dozens of ways, wouldn't it make sense to uplift the `test_package` to be an on-install/on-build feature too?

**Comment by memsharded** [comment]
The test_package runs:

- When you run ``conan create`` and there is a ``test_package`` folder there. Not on recipe creation (``export``), but on package creation.
- When you explicitly call ``conan test <folder test_package> <pkg-reference>``. This can be run for installed packages too, even for remote packages, and they will be installed in the local cache before running the test_package.

The only thing, is that they will run for the selected configuration, as most of other Conan commands (``install``, ``info``, ``create``, ``test``, ``export-pkg``...)

**Comment by Minimonium** [comment]
I understand where it _does_ run, I'm asking about the problem of it testing only the creator configuration, which is fairly basic considering that we allow changing a package layout basing on literally anything. There is an infinite number of configurations out there that we don't change. 
@Morwenn suggested in Slack that it could be beneficial to have `test_package` be exported with the recipe and check package state for a consumer's configuration.

**Comment by tjwrona1992** [comment]
Thanks for the clarification! I'm working on packaging up my first project and I'm still fairly new to figuring out all the Conan groundwork.

If I understand correctly, @Minimonium makes a pretty good point. It would be cool if the `test_package` stuff was automatically verifiable on the end of the consumer based on their specific configurations.

On that same point I would also like to have the build be able to automatically run my automated tests when the consumer tries to use my package to confirm everything works end-to-end.

**Comment by memsharded** [comment]
Oh, yes, we have discussed some times about embedding the "test_package" into the package itself. It had some challenges, and didn't prove to have a huge value, more like a nice to have. Running the ``test_package`` for each installed package would many times be very overkill, not intended for most of the cases, only for debugging purposes. And for debugging purposes, it is relatively easy to ``git clone`` and then ``conan test``.

> On that same point I would also like to have the build be able to automatically run my automated tests when the consumer tries to use my package to confirm everything works end-to-end.

That is the point, if the "automated tests" are unit tests that are build with the internals of the library, that cannot be run outside of the build itself, not with the final package. If they are "package checks", then yes, it might be possible.

I will label this for discussion and consideration again.

**Comment by KerstinKeller** [comment]
We've stumbled upon this too, and think that it would be beneficial if the `test_package` was part of the actual Conan package.

> Running the test_package for each installed package would many times be very overkill, not intended for most of the cases, only for debugging purposes.

It doesn't need to be run when consuming a package, just when building it (eg `--build missing`)

I see three main advantages of `test_package` belonging to the recipe:
- If you do a `--build outdated` the test package would not have to be build everytime, because Conan would know if it has changed or not. (Currently it is run always, even if the recipe is not outdated - we have to work around this in our CI)
- If a consumer builds with `--build missing` and for whatever reason their package cannot properly be consumed, then the consumer would get an error when building their dependency, instead of when building their application.
- Even more importantly, a package where `test_package` has failed can be uploaded to CI! it's in the cache, laying there, and Conan upload will pick it up. 

From my point it would be great if `test_package` where part of the conan package!

**Comment by memsharded** [comment]
We have considered this again a few times, but the core problem is that it requires changing server protocols and storage, which is not an easy task. 
We didn't have time to work on this, and we are very close to 2.0, this is a very large task, it is not possible that it will fit in scope, so moving it for 2.X consideration.

**Comment by KerstinKeller** [comment]
This is too bad 😞 
It's still a point which pains us from time to time. Happy if you keep in on the roadmap for 2.X.

**Comment by memsharded** [comment]
Following up on this.

Conan 2 got a new feature called "metadata files": https://docs.conan.io/2/devops/metadata.html

Note that page contains a section with an example using the metadata files as a mechanism to store the ``test_package``: https://docs.conan.io/2/devops/metadata.html#test-package-as-metadata

Closing this ticket as solved, please create new tickets as necessary for any further question or issue. Thanks for your feedback!
